src/GraphQL/Types/Collections/SchemaTypes.cs

```diff
@@ -82,10 +82,34 @@ public class SchemaTypes : IEnumerable<IGraphType>
         private readonly TypeCollectionContext _context;
         private readonly INameConverter _nameConverter;
         private readonly object _lock = new object();
-        private bool _sealed;
 
-        private SchemaTypes(ISchema schema, List<(Type clrType, Type graphType)> typeMappings)
+        /// <summary>
+        /// Initializes a new instance with no types registered
+        /// </summary>
+        protected SchemaTypes()
         {
+        }
+
+        /// <summary>
+        /// Initializes a new instance for the specified schema, and with the specified type resolver.
+        /// </summary>
+        /// <param name="schema">A schema for which this instance is created.</param>
+        /// <param name="graphTypeProvider">A service provider used to resolve graph types.</param>
+        public SchemaTypes(ISchema schema, IServiceProvider graphTypeProvider)
+        {
+            if (schema == null)
+                throw new ArgumentNullException(nameof(schema));
+            if (graphTypeProvider == null)
+                throw new ArgumentNullException(nameof(graphTypeProvider));
+
+            var types = GetSchemaTypes(schema, graphTypeProvider);
+            var typeMappingsEnumerable = schema.TypeMappings ?? throw new ArgumentNullException(nameof(schema) + "." + nameof(ISchema.TypeMappings));
+            var typeMappings = typeMappingsEnumerable is List<(Type, Type)> typeMappingsList ? typeMappingsList : typeMappingsEnumerable.ToList();
+            var directives = schema.Directives ?? throw new ArgumentNullException(nameof(schema) + "." + nameof(ISchema.Directives));
+            Func<Type, IGraphType> resolveType = t => (IGraphType)graphTypeProvider.GetRequiredService(t);
```

Comment:
Create old argument values from simplified constructor syntax, so rest of code does not need to change.